src/os/bluestore/BlueStore.cc

```diff
@@ -3108,6 +3109,23 @@ unsigned BlueStore::ExtentMap::decode_some(bufferlist& bl)
   return num;
 }
 
+//-------------------------------------------------------------------------
+void BlueStore::ExtentMap::add_shard_info_to_onode(bufferlist v, int shard_id)
```

Comment:
IMO all you need in this func is to call decode_some, which will load extents and blobs to extent map. You're not goind to use this onode later anyway...

void BlueStore::ExtentMap::add_shard_info_to_onode(bufferlist v)
{
decode_some(v);
}

Reply:
Or even make the next step forward... decode_some builds a temporary array of blobs this onode uses. Hence you can refactor it (or make a clone) which will load such an array for you. As a result you wouldn't need all the stuff from to build list of unique physical extents which is currently performed in read_allocation_from_single_onode

Reply:
Generally one should distinguish 3 types of blobs:
1) regular blobs which are attached to a specific onode only (i.e. aren't shared among multiple onodes) and serialized with the first referencing extent. decode_some is good in loading them in an unique set (aka vector<BlobRef> blobs).
2) spanning blobs - similarly to the above aren't shared but they're serialized separately from the extent map.  Not sure you're loading them at all...
3) shared blobs - blobs which are shared among multiple onodes (due to onode cloning) and hence serialized separately. I need to double check where(when) physical extents are serialized (IIRC that happens out of onode (de)serialization). But the key point is that one might find such blobs while reading multiple onodes - hence there is a need to handle that properly, i.e. avoid pextent duplications. I presume this issue is hidden with that sorted_extents_t structure you maintain but in fact it isn't enough since generally there is no guarantee that all the onodes referencing the same shared blob put their allocations into a single sorted_extents_t instance(or bunch/commit - don't know how to name that portion which is committed when one reaches memory cap).

Reply:
The are 3 places handling duplicated p_extents:
First, at read_allocation_from_single_onode() we store the extents in a temporary map and skipping duplication.
Second, the sorted_extent_t structure attempts to remove duplication, but extremely large allocations (i.e. more than 300M physical extents after the first filter) will not fit in a single batch and we will still pass duplication to the allocator.
Out last line of defense as I see it is the allocator itself which I assume can handle duplication.

Are you saying that the allocator can't handle duplication?
I can add another pass after building the allocator to read all its entries - sort, merge and remove duplicates.

Reply:
The spanning blobs comment is worrisome as that means I will break the allocation state in recovery flow.
Can you please schedule a meeting with me, you and Adam to get to the bottom of this?

Reply:
300M physical extents after the first filter) will not fit in a single batch and we will still pass duplication to the allocator.
> Out last line of defense as I see it is the allocator itself which I assume can handle duplication.
> 
> Are you saying that the allocator can't handle duplication?
No it's not mandatory for allocators to handle them (some implementations might assert though) - that's an additional performance overhead and useless when using free list manager..